ndk/src/media/media_codec.rs

```diff
@@ -245,9 +305,127 @@ impl MediaCodec {
             inner: NonNull::new(unsafe {
                 ffi::AMediaCodec_createEncoderByType(c_string.as_ptr())
             })?,
+            async_notify_callback: None,
         })
     }
 
+    #[cfg(feature = "api-level-28")]
+    pub fn set_async_notify_callback(
+        &mut self,
+        callback: Option<AsyncNotifyCallback>,
+    ) -> Result<()> {
+        unsafe extern "C" fn ffi_on_input_available(
+            _codec: *mut ffi::AMediaCodec,
+            user_data: *mut c_void,
+            index: i32,
+        ) {
+            abort_on_panic(|| {
+                let callback = &mut *(user_data as *mut AsyncNotifyCallback);
+                if let Some(f) = callback.on_input_available.as_mut() {
+                    f(index as usize);
+                }
+            })
+        }
+
+        unsafe extern "C" fn ffi_on_output_available(
+            _codec: *mut ffi::AMediaCodec,
+            user_data: *mut c_void,
+            index: i32,
+            buffer_info: *mut ffi::AMediaCodecBufferInfo,
+        ) {
+            abort_on_panic(|| {
+                let callback = &mut *(user_data as *mut AsyncNotifyCallback);
+                if let Some(f) = callback.on_output_available.as_mut() {
+                    let buffer_info = BufferInfo {
+                        inner: *buffer_info,
+                    };
+                    f(index as usize, &buffer_info);
+                }
+            })
+        }
+
+        unsafe extern "C" fn ffi_on_format_changed(
+            _codec: *mut ffi::AMediaCodec,
+            user_data: *mut c_void,
+            format: *mut ffi::AMediaFormat,
+        ) {
+            abort_on_panic(|| {
+                // Ownership of the format is not documented, but the implementation allocates a new instance and does
+                // not free it, so assume it is ok for us to do so
+                // https://cs.android.com/android/platform/superproject/main/+/refs/heads/main:frameworks/av/media/ndk/NdkMediaCodec.cpp;l=254;drc=5e15c3e22f3fa32d64e57302201123ce41589adf
+                let format = MediaFormat::from_ptr(NonNull::new_unchecked(format));
```

Comment:
I think the comment above mentions that it indeed makes a copy specifically for the client to optionally modify this type, let's include that in the link:


https://cs.android.com/android/platform/superproject/main/+/refs/heads/main:frameworks/av/media/ndk/NdkMediaCodec.cpp;l=248-254;drc=5e15c3e22f3fa32d64e57302201123ce41589adf

Reply:
Seems more like a safety thing rather than specifically being for the benefit of the client. In principle we could pass ownership of the `MediaFormat` to the callback instead of passing by reference, but given the ownership is not documented I don't think this is a very good idea. Acquiring ownership in the way we do is kind of a necessary evil to avoid a memory leak.

At any rate, I have updated the link.

Reply:
I won't be opposed to moving it, this is one of those implicit API things that Google cannot change anymore since everyone is expecting this to be a memory leak when they don't "own" the pointer (which is silly, but alas).

Also just noticing that all the `fn set_*()` functions take `&self` _im_mutable :eyes:

Reply:
I feel like most of the functions in this file probably ought to be `&mut self`. I'm not sure it makes sense to be allowing such free reign on a `MediaCodec` via a shared reference.

Reply:
We can still fix that before the next release, if you want @spencercw. CC @zarik5 in case there was a thoughtful reason for this?

Reply:
@zarik5 feel free to contribute that!

Also, are you saying that `MediaCodec` could perhaps be `Send`/`Sync`, does the NDK implement synchronization internally?